drivers/entropy/entropy_nrf5.c

```diff
@@ -296,6 +296,7 @@ static int entropy_nrf5_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len
 
 			buf[--len] = nrf_rng_random_value_get();
 			nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
+			NVIC_ClearPendingIRQ(RNG_IRQn);
```

Comment:
Wondering if this is a fix, the pending IRQ is cleared at line 308 before enabling the RNG_IRQn.

Reply:
Cortex-M4 Devices Generic User Guide:
>**Wakeup from WFE**
>The processor wakes up if:
>- it detects an exception with sufficient priority to cause exception entry
>- it detects an external event signal, see The external event input
>- in a multiprocessor system, another processor in the system executes an SEV instruction.
>
>In addition, if the SEVONPEND bit in the SCR is set to 1, any new pending interrupt triggers an event and wakes up the processor, even if the interrupt is disabled or has insufficient priority to cause exception entry. For more information about the SCR see System Control Register.

In our case the SEVONPEND bit in the SCR is set (by `_CpuIdleInit()` called from `kernel_arch_init()`) and we are waiting for disabled interrupt.

As only "*new pending interrupt triggers an event and wakes up the processor*", the WFE called in second run of the loop will block forever (assuming no other activity in SoC) as there will be no **new** pending interrupt.

Clearing the pending state of the interrupt solves this problem as system sees **new** pending interrupt during each run of the loop.

BTW: The same information about root cause and proposed solution of this problem is written in commit message.